### PR TITLE
Issue 22445 experiment targeting

### LINKED ISSUE #22445: Experiment Targeting setup
* Will be implemented using rules 
* We need to specify the experiment as parent of the rule. 
We can use https://demo.dotcms.com/api/v1/sites/a9f30020-54ef-494e-92ed-645e757171c2/ruleengine/rules **OR** create a new one
* It will be only one rule which you can add different conditions to. 
* The action/actionlet will be fixed. It will be a fixed experiment action that needs to be created
* It will become activated/enabled once the experiment is published/started

- [ ] Create experiment actionlet

- [ ] Add targeting condition

  ```
  POST /experiments/{key}/targetingConditions
  {  
  }
  ```

- [ ] Edit targeting condition
  ```
  PUT /experiments/{key}/targetingConditions/{id}
  {  
  }
  ```
- [ ] Delete targeting condition
  ```
  DELETE /experiments/{key}/targetingConditions/{id}
  ```

### PR DESCRIPTION
**Pull request by dsilvam**
Includes the ability to add, edit, or delete Targeting Conditions to Experiments via REST. 
It uses existing Rules under the covers. 
It basically creates one rule per experiment with one condition group and N conditions. 
A TargetingCondition immutable was created in order to abstract what's needed from the Experiments point of view. 
Includes postman tests.

### CODE REVIEW

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 1]
```diff
@@ -95,9 +107,91 @@ public Experiment save(final Experiment experiment, final User user) throws
...
+        Rule experimentRule;
+
+        if(UtilMethods.isSet(rules)) {
+            experimentRule = rules.get(0);
+        } else {
+            experimentRule = createRuleAndConditionGroup(experiment, user);
+        }
+
+        // transform and save TargetingConditions into conditions
+        experiment.targetingConditions().get().forEach((targetingCondition -> {
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:146]: Unnecessary pair of parenthesis.

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 2]
```diff
@@ -95,9 +107,91 @@ public Experiment save(final Experiment experiment, final User user) throws
...
+            createAndSaveCondition(user, experimentRule, targetingCondition);
+        }));
+
+    }
+
+    private void createAndSaveCondition(User user, Rule experimentRule,
+            TargetingCondition targetingCondition) {
+        Condition condition = targetingCondition.id().isPresent()
+            ? Try.of(()->rulesAPI.getConditionById(targetingCondition.id().get(), user, false))
```
Comment (victoralfaro-dotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:155]: Probably cleaner:
```java
Condition condition = targetingCondition.id()
            .map(id -> Try.of(()->rulesAPI.getConditionById(id, user, false))
                            .getOrElseThrow(()->new IllegalArgumentException("Invalid targeting Condition Id provided. Id: " + id))).orElse(createCondition(experimentRule, targetingCondition));
```

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 3]
```diff
@@ -95,9 +107,91 @@ public Experiment save(final Experiment experiment, final User user) throws
...
+    }
+
+    private Condition createCondition(final Rule experimentRule,
+            final TargetingCondition targetingCondition) {
+        final Condition condition = new Condition();
+        condition.setConditionGroup(experimentRule.getGroups().get(0).getId());
+        return condition;
+    }
+
+    @NotNull
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:178]: remove it

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java [hunk 4]
```diff
@@ -332,6 +450,27 @@ public Experiment deleteVariant(String experimentId, String variantName, User us
 
     }
 
+    @Override
+    public Experiment deleteTargetingCondition(String experimentId, String conditionId, User user)
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:454]: missing annotation?

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java (hunk 1 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:146]: Done

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java (hunk 3 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:178]: Done

File: dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java (hunk 2 above)
Comment (dsilvam) [reply dotCMS/src/main/java/com/dotcms/experiments/business/ExperimentsAPIImpl.java:155]: Gonna leave like it is so it's clear that it is based on the optional being present or not

**Review by victoralfaro-dotcms — APPROVED** [review]

**Review by jdotcms — APPROVED** [review]

**Pull request merged**